src/ocean_emulators/utils/location.py

```diff
@@ -0,0 +1,132 @@
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Annotated, Any, Literal, Self
+from urllib.parse import quote, urljoin, urlparse
+
+import xarray as xr
+from pydantic import BaseModel, BeforeValidator, model_serializer, model_validator
+
+
+class UnresolvedLocation(BaseModel):
+    """Representation for a raw string in a Location config.
+
+    This is expected to be a relative or absolute path.
+    It is flexibily interepreted as a url or local path depending
+    on what kind of location it is resolved against.
+    """
+
+    path: str
+
+    @model_validator(mode="after")
+    def validate_path(self) -> Self:
+        if urlparse(self.path).scheme:
+            raise ValueError(
+                "Absolute urls are not supported, please use a "
+                "relative path or set type = 's3' or 'local'"
+            )
+        return self
+
+    @model_serializer
+    def seralize(self) -> Any:
+        return self.path
+
+
+class ResolvedLocation(ABC):
+    """A location which is ready to be opened or resolved against."""
+
+    @abstractmethod
+    def open(self, chunks: dict[str, int] | None = None) -> xr.Dataset:
+        pass
+
+    @abstractmethod
+    def resolve(self, location: "Location") -> "ResolvedLocation":
```

Comment:
🐑  Does this work here?
```suggestion
    def resolve(self, location: "Location") -> Self:
```

Reply:
Yeah, I almost did that but we can't promise that resolve returns the class of `self` -- if you have `self: LocalLocation` and `location: S3Location` then `self.resolve(location)` returns a S3Location, not a LocalLocation (in this case, Self would resolve to a LocalLocation). All we can promise is that it's *some* `ResolvedLocation`, not necessarily the same type as the instance you're calling it on.

Reply:
(Sorry, I ninja-edited my comment when I went to actually go do this and realized we can't)